### PR TITLE
fix(webhook): Move all the webhook response processing out of task ex…

### PR DESCRIPTION
**Pull request by srekapalli**
- No functionality change.
- Moved all the response and exception processing out of the task class.

### CODE REVIEW

**Review by marchello2000 — APPROVED** [review]

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+import java.util.stream.IntStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.util.StringUtils;
+import org.springframework.web.client.HttpStatusCodeException;
+
+/** processes response or the client exception received from a webhook call. */
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:41]: I'm not totally clear on this doc. Perhaps:

```suggestion
/** Processes responses and errors from webhook stage executions. */
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+        return processClientException(exceptionReceived);
+      }
+    }
+
+    // Process received response.
+    if (response != null) {
+      return processResponse(response);
+    }
+
+    throw new IllegalArgumentException("No response or exception to process.");
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:77]: While we're in here, would like to see usage of `SpinnakerException` family.

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  TaskResult processReceivedFailureStatusCode(HttpStatus status, Map<String, Object> stageOutput) {
+    String errorMessage;
+    ExecutionStatus executionStatus;
+    Map<String, Object> webHookOutput = (Map<String, Object>) stageOutput.get("webhook");
+    // Fail fast status check, retry status check or fail permanently
+    if (stageData.failFastStatusCodes != null
+        && stageData.failFastStatusCodes.contains(status.value())) {
+      errorMessage =
+          String.format(
+              "Received a status code %s configured to fail fast, terminating stage for pipeline %s to %s",
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:102]: While we're in here...

```suggestion
              "Received status code %s, which is configured to fail fast, terminating stage for pipeline %s to %s",
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+    return TaskResult.builder(executionStatus).context(stageOutput).build();
+  }
+
+  TaskResult processClientException(Exception e) {
+    String errorMessage;
+    ExecutionStatus executionStatus;
+    if (e instanceof UnknownHostException || e.getCause() instanceof UnknownHostException) {
+      errorMessage =
+          String.format(
+              "Name resolution failure in webhook for pipeline %s to %s, will retry.",
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:131]: While we're in here...

```suggestion
              "Remote host resolution failure in webhook for pipeline %s to %s, will retry.",
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            || e.getCause() instanceof SocketTimeoutException)) {
+      errorMessage =
+          String.format(
+              "Socket timeout in webhook on GET request for pipeline %s to %s, will retry.",
+              executionId, stageData.url);
+      executionStatus = ExecutionStatus.RUNNING;
+    } else {
+      errorMessage =
+          String.format(
+              "An exception occurred for pipeline %s to %s. %s",
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:145]: While we're in here...

```suggestion
              "An exception occurred for pipeline %s performing a request to %s. %s",
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+          webHookOutput.put(
+              "error", "Exception while resolving status check URL: " + e.getMessage());
+          log.error("Exception received while determining status check url", e);
+          return TaskResult.builder(ExecutionStatus.TERMINAL).context(stageOutput).build();
+        }
+
+        if (StringUtils.isEmpty(statusUrl)) {
+          webHookOutput.put(
+              "error",
+              "The status URL couldn't be resolved, but 'Wait for completion' was checked");
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:184]:
```suggestion
              "The status URL couldn't be resolved, but 'Wait for completion' is configured");
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+      }
+
+      // Retrieve artifacts
+      if (stageExecution.getContext().containsKey("expectedArtifacts")
+          && !((List) stageExecution.getContext().get("expectedArtifacts")).isEmpty()) {
+        try {
+          stageOutput.put("artifacts", JsonPath.parse(response.getBody()).read("artifacts"));
+        } catch (Exception e) {
+          webHookOutput.put(
+              "error", "Expected artifacts in webhook response couldn't be parsed " + e.toString());
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:200]:
```suggestion
              "error", "Expected artifacts in webhook response couldn't be parsed: " + e.toString());
```

File: orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+    try {
+      if (responseBody.startsWith("{")) {
+        body = objectMapper.readValue(responseBody, Map.class);
+      } else if (responseBody.startsWith("[")) {
+        body = objectMapper.readValue(responseBody, List.class);
+      }
+    } catch (JsonProcessingException ex) {
+      // Just leave body as string, probs not JSON
+      log.warn("Failed to parse webhook payload as JSON", ex);
+    }
```
Comment (robzienert) [inline orca-webhook/src/main/groovy/com/netflix/spinnaker/orca/webhook/tasks/WebhookResponseProcessor.java:222]: Don't have to fix it here, but this behavior is incorrect: It assumes that a valid JSON response must only be either a list or object, however according to the JSON spec, the root element can be of any of JSON's 7 types: https://www.json.org/json-en.html

**Review by robzienert — APPROVED** [review]
Took this as an opportunity to audit the error messages we send to users.

Also noticed a bug (pre-existing). Added an issue to track, maybe community can pick it up... seems low-pri. https://github.com/spinnaker/spinnaker/issues/6040

**Pull request merged**